modules/gapi/samples/onevpl_infer_single_roi.cpp

```diff
@@ -0,0 +1,254 @@
+            const float rc_bottom  = data[i * OBJECT_SIZE + 6];
+            (void) label; // unused
+
+            if (image_id < 0.f) {
+                break;    // marks end-of-detections
+            }
+            if (confidence < 0.5f) {
+                continue; // skip objects with low confidence
+            }
+
+            // map relative coordinates to the original image scale
+            // taking the ROI into account
+            cv::Rect rc;
+            rc.x      = static_cast<int>(rc_left   * up_roi.width);
+            rc.y      = static_cast<int>(rc_top    * up_roi.height);
+            rc.width  = static_cast<int>(rc_right  * up_roi.width)  - rc.x;
+            rc.height = static_cast<int>(rc_bottom * up_roi.height) - rc.y;
+            rc.x += in_roi.x;
+            rc.y += in_roi.y;
+            out_objects.emplace_back(rc & surface);
+        }
+    }
+};
+
+GAPI_OCV_KERNEL(OCVBBoxes, BBoxes) {
+    // This kernel converts the rectangles into G-API's
+    // rendering primitives
+    static void run(const std::vector<cv::Rect> &in_face_rcs,
+                    const             cv::Rect  &in_roi,
+                          std::vector<cv::gapi::wip::draw::Prim> &out_prims) {
+        out_prims.clear();
+        const auto cvt = [](const cv::Rect &rc, const cv::Scalar &clr) {
+            return cv::gapi::wip::draw::Rect(rc, clr, 2);
+        };
+        out_prims.emplace_back(cvt(in_roi, CV_RGB(0,255,255))); // cyan
+        for (auto &&rc : in_face_rcs) {
+            out_prims.emplace_back(cvt(rc, CV_RGB(0,255,0)));   // green
+        }
+    }
+};
+
+} // namespace custom
+
+int main(int argc, char *argv[]) {
+
+    cv::CommandLineParser cmd(argc, argv, keys);
+    cmd.about(about);
+    if (cmd.has("help")) {
+        cmd.printMessage();
+        return 0;
+    }
+
+    // get file name
+    std::string file_path = cmd.get<std::string>("input");
+    const std::string output = cmd.get<std::string>("output");
+    const auto face_model_path = cmd.get<std::string>("facem");
+
+    // check ouput file extension
+    if (!output.empty()) {
+        auto ext = output.find_last_of(".");
+        if (ext == std::string::npos || (output.substr(ext + 1) != "avi")) {
+            std::cerr << "Output file should have *.avi extension for output video" << std::endl;
+            return -1;
+        }
+    }
+
+    auto face_net = cv::gapi::ie::Params<custom::FaceDetector> {
+        face_model_path,                 // path to topology IR
+        get_weights_path(face_model_path),   // path to weights
+        cmd.get<std::string>("faced"),   // device specifier
+    };
+    auto kernels = cv::gapi::kernels
+        < custom::OCVLocateROI
+        , custom::OCVParseSSD
+        , custom::OCVBBoxes>();
+    auto networks = cv::gapi::networks(face_net);
+
+    // Create source
+    cv::Ptr<cv::gapi::wip::IStreamSource> cap;
+    try {
+        cap = cv::gapi::wip::make_vpl_src(file_path);
```

Comment:
Could we implement that through "Kernel packages"?
Their definition can be header-only, but implementation requires extra dependencies, or been dynamically loadable.

/cc @dmatveev

Reply:
@alalek For media sources, it is not clear yet.

Sure we can move the decode to the graph level so it becomes an "Operation" (with a single-header definition), and its implementations become "Kernels", so VPL or GStreamer could be specified via a kernel package -- and this mechanism would work there for free.

This approach has its cons, though, so I doubt if decode should be a graph operation or be a graph source. Let it (and GStreamer) be sources for now -- and for sources, probably sooner or later we'll introduce something factory-like to construct them based on capabilities, options, and so on